src/providers/completer/citation.ts

```diff
@@ -1,137 +1,67 @@
 import * as vscode from 'vscode'
 import * as fs from 'fs'
+import {bibtexParser} from 'latex-utensils'
 
 import {Extension} from '../../main'
 
-const bibEntries = ['article', 'book', 'bookinbook', 'booklet', 'collection', 'conference', 'inbook',
-                    'incollection', 'inproceedings', 'inreference', 'manual', 'mastersthesis', 'misc',
-                    'mvbook', 'mvcollection', 'mvproceedings', 'mvreference', 'online', 'patent', 'periodical',
-                    'phdthesis', 'proceedings', 'reference', 'report', 'set', 'suppbook', 'suppcollection',
-                    'suppperiodical', 'techreport', 'thesis', 'unpublished']
-
-interface CitationRecord {
+export interface Suggestion extends vscode.CompletionItem {
     key: string,
-    [key: string]: string | undefined
+    detail: string,
+    fields: {[key: string]: string},
+    file: string,
+    position: vscode.Position
 }
 
 export class Citation {
     extension: Extension
-    suggestions: vscode.CompletionItem[]
-    citationInBib: { [id: string]: {citations: CitationRecord[], rootFiles: string[] }} = {}
```

Comment:
citations used to be tight to a list of LaTeX projects via `rootFiles` making it possible to have different citation completions depending on the active editor without forgetting/reparsing `.bib` files every time the time changes (see #1085). How is it done now?

Reply:
I missed that 🤣 The latest commit re-implement the feature. Here is a sketch:

First, the included bib file information is now stored in `cachedContent`. Additionally, these bib files are watched by watcher, which updates the entries upon changes.

When the extension provides citation suggestions (inline or browser), it first traces from the root file and investigate all included bib files along the parent-child-tree. Only those included will be used in the suggestion, along with inline `bibitem`s.